File: PerudoBot.Database/Data/Action.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace PerudoBot.Database.Data
{
    public abstract class Action : TrackedEntity
    {
        public int Id { get; set; }
        public string ActionType { get; set; }

        [ForeignKey("RoundId")]
        public Round Round { get; set; }
        public int RoundId { get; set; }


        [ForeignKey("PlayerId")]
        public Player Player { get; set; }
        public int PlayerId { get; set; }


        [ForeignKey("PlayerHandId")]
        public PlayerHand PlayerHand { get; set; }
        public int? PlayerHandId { get; set; }

        [ForeignKey("ParentActionId")]
        public Action ParentAction { get; set; }
        public int? ParentActionId { get; set; }
    }

    public class BidAction : Action
    {
        public int Quantity { get; set; }
        public int Pips { get; set; }
    }

    public class LiarAction : Action
    {
        public int LosingPlayerId { get; set; }
        public int LivesLost { get; set; }
        public int ActualQuantity { get; set; }

        [NotMapped]
        public bool IsSuccessful => LosingPlayerId != PlayerId;

        [NotMapped]
        public BidAction TargetBid => Round.LatestBid;
    }

    public class BetAction : Action
    {
        public int BetAmount { get; set; }
        public int BetType { get; set; }

        public double BetOdds { get; set; }

        public bool IsSuccessful { get; set; }

        [ForeignKey("TargetBidId")]
        public BidAction TargetBid { get; set; }

        public int TargetBidId { get; set; }

        [NotMapped]
        public int BetQuantity => TargetBid.Quantity;

        [NotMapped]
        public int BetPips => TargetBid.Pips;
    }
}

File: PerudoBot.Discord/cogs/models/action.py
class Action():
    def __init__(self, json: dict):
        if json is None: return
        self.player_id = json.get('playerId')
        self.action_type = json.get('actionType')

class Bid(Action):
    def __init__(self, json: dict):
        super().__init__(json)
        self.quantity = json.get('quantity')
        self.pips = json.get('pips')

class Liar(Action):
    def __init__(self, json: dict):
        super().__init__(json)
        self.target_bid : Bid = Bid(json.get('targetBid'))
        self.lives_lost = json.get('livesLost')
        self.losing_player_id = json.get('losingPlayerId')
        self.is_successful = json.get('isSuccessful')
        self.actual_quantity = json.get('actualQuantity')

class Bet(Action):
    def __init__(self, json: dict) -> None:
        super().__init__(json)
        self.target_bid : Bid = Bid(json.get('targetBid'))
        self.bet_type = json.get('betType')
        self.bet_amount = json.get('betAmount')
        self.is_successful = json.get('isSuccessful')
        self.bet_odds = json.get('betOdds')